### PR TITLE
Fix incorrect redirectUrl check with external authentication (#14198)

### LINKED ISSUE #14198: Umbraco doesn't honor the "returnPath" after external login
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

11.3.0

### Bug summary

If my backoffice session is outdated I am (transparently) redirected to my external login page (Azure AD in this case) and then back to the CMS start page - not to the URL specified in the `returnPath` of the query string.

### Specifics

image

### Steps to reproduce

* Setup an external login system for the backoffice (e.g. Azure AD)
* Make sure you are logged OUT of the backoffice
* Try to access an internal Umbraco page e.g. `https://cms-stage.lrf.se/umbraco#/settings`
* Notice you are asked to login, with the returnPath being the value of the URL you are trying to access
* Proceed to log in
* Notice you are being redirected back to the Umbraco start page, not the URL specified

### Expected result / actual result

I expect to be redirected to the value in the `returnPath` query parameter after logging in. Instead, I'm being redirected to the backoffice start page.

### PR DESCRIPTION
**Pull request by jrunestone**
### Prerequisites

Fixes #14198 

### Description
When using an external authentication method to sign in to Umbraco back office the `returnUrl` is checked if it's absolute and if so (disallowed) resets it to the start page. The check was incorrect though and resulted in every redirectUrl being reset so the redirect url was never honored. (`Uri.TryCreate` would create an absolute `file://` uri out of anything it got).

NOTE: I would use `IUrlHelper.IsLocalUrl` but it's not available to me and the only difference is it will allow `~/` as well, and it's not relevant here.

### CODE REVIEW

**Comment by emmagarland** [comment]
Hi @jrunestone 

Thanks for your PR to fix #14198, where Umbraco doesn't honour the return path after external login. 

One of the Core Collaborators team will review this as soon as possible.

Best, 

Emma

File: src/Umbraco.Web.BackOffice/Controllers/BackOfficeController.cs [hunk 1]
```diff
@@ -329,7 +329,7 @@ public async Task<JavaScriptResult> ServerVariables()
     [AllowAnonymous]
     public ActionResult ExternalLogin(string provider, string? redirectUrl = null)
     {
-        if (redirectUrl == null || Uri.TryCreate(redirectUrl, UriKind.Absolute, out _))
+        if (redirectUrl == null || !redirectUrl.StartsWith("/"))
```
Comment (lauraneto) [inline src/Umbraco.Web.BackOffice/Controllers/BackOfficeController.cs:332]: This was a bit of a tricky one. 😅
I was trying to reproduce it in my Windows laptop and it took me a while to realize that the issue was only present in Linux/macOS.
Found this dotnet issue about it: https://github.com/dotnet/runtime/issues/22718
```suggestion
        // Only relative urls are accepted as redirect url.
        // We can't simply use Uri.TryCreate with kind Absolute, as in Linux any relative url would be seen as an absolute file uri
        if (redirectUrl == null || !Uri.TryCreate(redirectUrl, UriKind.RelativeOrAbsolute, out Uri? redirectUri) || redirectUri.IsAbsoluteUri)
```
^ I do believe that this suggestion is better, but let me know what you think! 🙂

**Comment by lauraneto** [comment]
Hi @jrunestone ,

I went ahead and applied my suggestion and merged the PR. I hope that's ok!
This fix will be included in 12.2.
Thanks again for figuring out the issue and making a pull request!! 🎉

**Comment by jrunestone** [comment]
Hi @lauraneto, Sorry for the late response (vacation times). Thanks for the update and for merging! :) :partying_face:

**Pull request merged**